containerd-shim-wws-v1/src/executor.rs

```diff
@@ -17,65 +25,75 @@ const WWS_ADDR: &str = "0.0.0.0";
-            ExecutorError::Other(format!("failed to prepare stdio for container: {}", err))
-        })?;
+            prepare_stdio(self.stderr).map_err(|err| {
+                ExecutorError::Other(format!("failed to prepare stdio for container: {}", err))
+            })?;
 
-        let path = PathBuf::from("/");
+            let path = PathBuf::from("/");
 
-        let config = match Config::load(&path) {
-            Ok(c) => c,
-            Err(err) => {
-                error!("[wws] There was an error reading the .wws.toml file. It will be ignored");
-                error!("[wws] Error: {err}");
+            let config = match Config::load(&path) {
+                Ok(c) => c,
+                Err(err) => {
+                    error!(
+                        "[wws] There was an error reading the .wws.toml file. It will be ignored"
+                    );
+                    error!("[wws] Error: {err}");
 
-                Config::default()
-            }
-        };
+                    Config::default()
+                }
+            };
 
-        // Check if there're missing runtimes
-        if config.is_missing_any_runtime(&path) {
-            error!("[wws] Required language runtimes are not installed. Some files may not be considered workers");
-            error!("[wws] You can install the missing runtimes with: wws runtimes install");
-        }
+            // Check if there're missing runtimes
+            if config.is_missing_any_runtime(&path) {
+                error!("[wws] Required language runtimes are not installed. Some files may not be considered workers");
+                error!("[wws] You can install the missing runtimes with: wws runtimes install");
+            }
 
-        let routes = Routes::new(&path, "", Vec::new(), &config);
-
-        let rt = Runtime::new().unwrap();
-        rt.block_on(async {
-            let f = serve(&path, routes, WWS_ADDR, WWS_PORT, false, None)
-                .await
-                .unwrap();
-            info!(" >>> notifying main thread we are about to start");
-            tokio::select! {
-                _ = f => {
-                    log::info!(" >>> server shut down: exiting");
-                    std::process::exit(0);
-                },
-            };
-        });
-        std::process::exit(137);
-    }
+            let routes = Routes::new(&path, "", Vec::new(), &config);
 
-    fn can_handle(&self, _spec: &Spec) -> bool {
-        true
+            let rt = Runtime::new().unwrap();
+            rt.block_on(async {
+                let f = serve(&path, routes, WWS_ADDR, WWS_PORT, false, None)
+                    .await
+                    .unwrap();
+                info!(" >>> notifying main thread we are about to start");
+                tokio::select! {
+                    _ = f => {
+                        log::info!(" >>> server shut down: exiting");
+                        std::process::exit(0);
+                    },
+                };
```

Comment:
**fixe:** you could use await here, right?
**suggestion:**

```rust
#[derive(Clone)]
pub struct WwsExecutor {
    pub stdio: Stdio,
}

impl WwsExecutor {
    pub fn new(stdio: Stdio) -> Self {
        Self { stdio }
    }
}

impl Executor for WwsExecutor {
    fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
        if is_linux_executable(spec).is_ok() {
            log::info!("executing linux container");
            LinuxContainerExecutor::new(self.stdio.take()).exec(spec)
        } else {
            if let Err(err) = self.wasm_exec(spec) {
                log::info!(" >>> server shut down due to error: {err}");
                std::process::exit(137);
            }
            log::info!(" >>> server shut down: exiting");
            std::process::exit(0);
        }
    }

    fn validate(&self, _spec: &Spec) -> Result<(), ExecutorValidationError> {
        Ok(())
    }
}

impl WwsExecutor {
    fn wasm_exec(&self, _spec: &Spec) -> Result<()> {
        let stderr = self.stdio.take().stderr;
        stderr.redirect().context("redirecting stdio")?;

        let root = Path::new("/");

        let config = Config::load(root).unwrap_or_else(|err| {
            error!("[wws] Error reading .wws.toml file. It will be ignored");
            error!("[wws] Error: {err}");
            Config::default()
        });

        // Check if there're missing runtimes
        if config.is_missing_any_runtime(root) {
            error!("[wws] Required language runtimes are not installed. Some files may not be considered workers");
            error!("[wws] You can install the missing runtimes with: wws runtimes install");
        }

        let routes = Routes::new(root, "", Vec::new(), &config);

        let rt = Runtime::new().context("failed to create runtime")?;
        rt.block_on(self.wasm_exec_async(root, routes))
    }

    async fn wasm_exec_async(&self, root: &Path, routes: Routes) -> Result<()> {
        let server = serve(root, routes, WWS_ADDR, WWS_PORT, false, None).await?;
        info!(" >>> notifying main thread we are about to start");
        Ok(server.await?)
    }
}
```